ValveResourceFormat/IO/NmClipExtract.cs

```diff
@@ -0,0 +1,221 @@
+            // The array below indexes into the bone sampling chain, which in turn indexes into the skeleton bones.
+            var modelSpaceBoneSamplingIndices = clip.Data.GetIntegerArray("m_modelSpaceBoneSamplingIndices");
+
+            var bonesToSampleInModelSpace = new KVObject("m_bonesToSampleInModelSpace", true, modelSpaceBoneSamplingIndices.Length);
+            foreach (var chainIdx in modelSpaceBoneSamplingIndices)
+            {
+                if (chainIdx < 0 || chainIdx >= modelSpaceSamplingChain.Length)
+                {
+                    throw new InvalidDataException($"Model space sampling chain index {chainIdx} is out of bounds (0..{modelSpaceSamplingChain.Length - 1}).");
+                }
+                var boneIdx = modelSpaceSamplingChain[chainIdx].GetInt32Property("m_nBoneIdx");
+                bonesToSampleInModelSpace.AddItem(skeleton.Bones[boneIdx].Name);
+            }
+            kv.AddProperty("m_bonesToSampleInModelSpace", bonesToSampleInModelSpace);
+
+            contentFile.AddSubFile(sourceFileName, () =>
+            {
+                return ModelExtract.ToDmxAnim(skeleton, [], animation);
+            });
+        }
+        var events = clip.Data.GetArray<KVObject>("m_events");
+        var docEventTracks = new KVObject("m_eventTracks", true, events.Length);
+        foreach (var ev in events)
+        {
+            var docEventTrack = BuildDocEventBasedOnEventClass(ev, ev.GetStringProperty("_class"));
+            var startTimeSeconds = ev.GetFloatProperty("m_flStartTimeSeconds");
+            var durationSeconds = ev.GetFloatProperty("m_flDurationSeconds");
+            // Doc file event time stamps are given in frames they can be technically floats, but based on recompilation tests
+            // these seem inconsistent, unless they're floored to int, then it matches up.
+            docEventTrack.GetArray<KVObject>("m_events").First().AddProperty("m_flStartTime", Math.Floor(startTimeSeconds * animation.Fps));
+            docEventTrack.GetArray<KVObject>("m_events").First().AddProperty("m_flDuration", Math.Floor(durationSeconds * animation.Fps));
+            docEventTracks.AddItem(docEventTrack);
+        }
+        kv.AddProperty("m_eventTracks", docEventTracks);
+        contentFile.Data = Encoding.UTF8.GetBytes(new KV3File(kv).ToString());
+        return contentFile;
+    }
+
+    // Returns a full event track.
+    private static KVObject BuildDocEventBasedOnEventClass(KVObject kvCompiledEvent, string className)
+    {
+        // From testing one event track in doc seems to correspond to one event in compiled asset
+        // even though m_events is an array inside each track.
+        var kvDocEventTrack = new KVObject(null);
+        var kvDocEvent = new KVObject("m_event");
+        kvDocEventTrack.AddProperty("m_type", "Duration");
+        kvDocEventTrack.AddProperty("m_bIsSyncTrack", kvCompiledEvent.ContainsKey("m_syncID"));
+        switch (className)
```

Comment:
I notice that most properties are directly copied from the source, what if all properties were copied, and only ones that need to be renamed were renamed?

Does this actually map all events, or are there more? Would be nice if we didn't have to manually maintain the full mapping.

Reply:
Yeah we do similar thing in other decompilers such as AnimationGraph.

You can extract class name by removing "CNm" and "Event", and create "CNmClipDocEvent_{class}". Pass through all properties by default, and only handle any special ones.